aws_lambda_builders/workflows/rust_cargo/actions.py

```diff
@@ -0,0 +1,216 @@
+    """
+    spec = handler.split(".", 1)
+    if len(spec) == 1:
+        return (spec[0], spec[0])
+    else:
+        return (spec[0], spec[1])
+
+
+class BuildAction(BaseAction):
+    NAME = "CargoBuild"
+    DESCRIPTION = "Building the project using Cargo"
+    PURPOSE = Purpose.COMPILE_SOURCE
+
+    def __init__(self, source_dir, handler, binaries, platform, mode, osutils=OSUtils()):
+        """
+        Build the a rust executable
+
+        :type source_dir: str
+        :param source_dir:
+            Path to a folder containing the source code
+
+        :type handler: str
+        :param handler:
+            Handler name in `package.bin_name` or `bin_name` format
+
+        :type binaries: dict
+        :param binaries:
+            Resolved path dependencies
+
+        :type platform: string
+        :param platform:
+            Platform builder is being run on
+
+        :type mode: str
+        :param mode:
+            Mode the build should produce
+
+        :type osutils: object
+        :param osutils:
+            Optional, External IO utils
+        """
+        self.source_dir = source_dir
+        self.handler = handler
+        self.mode = mode
+        self.binaries = binaries
+        self.platform = platform
+        self.osutils = osutils
+
+    def cargo_metadata(self):
+        p = self.osutils.popen(
+            ["cargo", "metadata", "--no-deps", "--format-version=1"],
+            stderr=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            env=os.environ.copy(),
+            cwd=self.source_dir,
+        )
+        out, err = p.communicate()
+        if p.returncode != 0:
+            raise BuilderError(message=err.decode("utf8").strip())
+        return json.loads(out)
+
+    def build_command(self, package):
+        cmd = [self.binaries["cargo"].binary_path, "build", "-p", package]
+        if self.mode != BuildMode.DEBUG:
+            cmd.append("--release")
+        if self.platform.lower() != "linux":
+            cmd.extend(["--target", "x86_64-unknown-linux-musl"])
+        return cmd
+
+    def resolve_binary(self, cargo_meta):
+        """
+        Interrogate cargo metadata to resolve a handler function
+
+        :type cargo_meta: dict
+        :param cargo_meta:
+            Build metadata emitted by cargo
+        """
+        (package, binary) = parse_handler(self.handler)
+        exists = any(
+            map(
+                lambda p: p["name"] == package
+                and any(map(lambda t: any(map(lambda k: k == "bin", t["kind"])) and t["name"] == binary, p["targets"])),
+                cargo_meta["packages"],
+            )
+        )
+        if not exists:
+            raise BuilderError(message="Cargo project does not contain a {handler} binary".format(handler=self.handler))
+
+        return (package, binary)
+
+    def build_env(self):
+        env = os.environ.copy()
+        if self.platform.lower() == "darwin":
```

Comment:
of note, there is likely a windows analog to this but I have not figured out what that looks like in practice and lack a proper windows env to test with